ucb-rit/coldfront
# Add support for purchasing additional SUs for Recharge projects

Fixes #291

**Changes**
- Added a "Purchase Service Units" button to the Project Detail view for Recharge (`ac_`) projects, accessible to superusers, PIs, and managers.
    - This button redirects to a landing page, from whence PIs and managers can submit a request to purchase more SUs (if there isn't already a pending request), as well as view pending and completed requests.
- Added a list view and detail view for these new requests; the detail view is readable by superusers, PIs, managers, and staff; and writable by superusers.
- Added views for reviewing a specific request, including:
    - Confirming that the MOU has been signed and that funds have been transferred, and
    - Denying a request for a given reason.
- Added a new model `AllocationAdditionRequest` to store these requests.
- Added tests.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes. The relevant new tests are:
    ```
    coldfront.core.project.tests.test_utils.test_addition_utils
    coldfront.core.project.tests.test_views.test_addition_views
    ```
- Manual Testing
    - Run the following commands (or simply re-run the Ansible playbook).
        ```
        python manage.py migrate allocation
        python manage.py add_allocation_defaults
        python manage.py create_staff_group
        ```
    - Turn on a debugging email server.
    - Ensure that requests can only be made for Recharge projects and only by their PIs or Managers.
    - Ensure that submitting a request sends an email.
    - Ensure that the relevant requests appear for the relevant users in the List and Detail views.
    - Ensure that approving/denying a request modifies database objects as expected and sends an email.
